rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,302 @@
+- Feature Name: `glides-trainsheet-events`
+- Start Date: 2023-01-20
+- RFC PR: [mbta/technology-docs#18](https://github.com/mbta/technology-docs/pull/18)
+- Status: Proposed
+
+# Summary
+Glides will provide high-level trainsheet information as CloudEvents in a Kinesis stream
+
+# Motivation
+As part of digitizing the currently paper-based trainsheets that define operational conditions on the Green Line, Glides will be gathering information about consist assignments, dropped/added trips, and departure time adjustments from its users: pull-out inspectors at selected terminal stations. This data can serve to improve the rider-facing predictions produced by RTR, especially for departure times at a terminals, which currently do not have departure predictions at all.
+
+# Guide-level explanation
+The Glides trainsheet functionality hews closely to what the paper-based trainsheets provide.
+
+- sign an operator in, marking them fit-for-duty
+- add a trip (1-way from the current terminal, 1-way arriving at the current terminal, or a round-trip from the current terminal)
+- drop a trip
+- add/change/remove train cars to a trip
+- add/change/remove the operator assigned to a trip
+- set a new departure time for a trip
+
+These actions are always taken by a logged-in user (generally an inspector).
+
+## CloudEvents
+All events will be in the [CloudEvents v1.0.2](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md) format (or later), using the JSON encoding. The event types will be under the `com.mbta.ctd.glides` event namespace, and will be in the past-tense (i.e. `trip-added` rather than `add-trip`).
+
+If multiple events happen as a part of the same user action, they can be written as a Batch event, wrapping all of the events in a JSON array.
+
+## Kinesis
+Events will be written to a Kinesis stream. Each Glides environment (`dev`, `dev-green`, and `prod`) will have a separate stream for their events. The partition key will be a hash of the station at which the inspector is working and the inspector's identity, which ensures that multiple events from a single inspector are ordered correctly if the messages are distributed across multiple shards.
+
+## Value types
+### Trip Key
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matches a GTFS trip, in which case RTR will us the GTFS trip ID in publishing the information.
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+| Added Trip | Scheduled Trip |
+| ---------- | -------------- |
+| `{"serviceDate": "2023-01-20", "id": "ADDED-123"}`|`{"serviceDate": "2023-01-20", "stop": "place-lake", "time": "01:02:30"}`|
```

Comment:
Good point on clarifying the type of the time, and the example.

In the paper timesheets, how are two trips which are scheduled to depart from a terminal at the same time distinguished?

Reply:
This happened for outbound trips starting at Government Center, where there are no paper trainsheets. At all of the branch terminals where there's a pullout inspector and trainsheets, times are unique.

Hypothetically if two trips that departed at the same time did appear at a terminal with trainsheets, they'd probably be distinguished by the "route" (AVI number).

Reply:
Since there isn't a pull-out inspector there, I don't think that's covered by this RFC. It also doesn't include route information, for example. 

If support in Glides isn't on your roadmap at the moment, I'll mention this as something to consider with future updates.

Reply:
We still need to uniquely refer to those trips because they do show up as arrivals in Glides and inspectors regularly make edits to them.

Reply:
I think with the new clarification (times are relative to the origin/destination), they should be unique.